crates/indexer/src/db.rs

```diff
@@ -0,0 +1,940 @@
+    ) -> Result<(), DbError>;
+
+    async fn add_deposit_event(
+        &self,
+        account: Address,
+        value: U256,
+        metadata: &TxMetadata,
+    ) -> Result<(), DbError>;
+
+    async fn add_withdrawal_event(
+        &self,
+        account: Address,
+        value: U256,
+        metadata: &TxMetadata,
+    ) -> Result<(), DbError>;
+
+    async fn add_stake_deposit_event(
+        &self,
+        account: Address,
+        value: U256,
+        metadata: &TxMetadata,
+    ) -> Result<(), DbError>;
+
+    async fn add_stake_withdrawal_event(
+        &self,
+        account: Address,
+        value: U256,
+        metadata: &TxMetadata,
+    ) -> Result<(), DbError>;
+
+    async fn add_callback_failed_event(
+        &self,
+        request_id: U256,
+        callback_address: Address,
+        error_data: Vec<u8>,
+        metadata: &TxMetadata,
+    ) -> Result<(), DbError>;
+}
+
+pub type DbObj = Arc<dyn IndexerDb + Send + Sync>;
+
+#[derive(Debug, Clone)]
+pub struct AnyDb {
+    pool: AnyPool,
+}
+
+impl AnyDb {
+    /// For SQLite use a `sqlite:file_path` URL; for Postgres `postgres://`.
+    pub async fn new(conn_str: &str) -> Result<Self, DbError> {
+        install_default_drivers();
+        let opts = AnyConnectOptions::from_str(conn_str)?;
+
+        let pool = AnyPoolOptions::new()
+            // you can tweak these per‐DB by inspecting opts.kind()
+            .max_connections(5)
+            .connect_with(opts)
+            .await?;
+
+        // apply any migrations
+        sqlx::migrate!().run(&pool).await?;
+
+        Ok(Self { pool })
+    }
+}
+
+#[async_trait]
+impl IndexerDb for AnyDb {
+    async fn get_last_block(&self) -> Result<Option<u64>, DbError> {
+        let res = sqlx::query("SELECT block FROM last_block WHERE id = $1")
+            .bind(SQL_BLOCK_KEY)
+            .fetch_optional(&self.pool)
+            .await?;
+
+        let Some(row) = res else {
+            return Ok(None);
+        };
+
+        let block_str: String = row.try_get("block")?;
+
+        Ok(Some(block_str.parse().map_err(|_err| DbError::BadBlockNumb(block_str))?))
+    }
+
+    async fn set_last_block(&self, block_numb: u64) -> Result<(), DbError> {
+        let res = sqlx::query(
+            "INSERT INTO last_block (id, block) VALUES ($1, $2)
+         ON CONFLICT (id) DO UPDATE SET block = EXCLUDED.block",
+        )
+        .bind(SQL_BLOCK_KEY)
+        .bind(block_numb.to_string())
+        .execute(&self.pool)
+        .await?;
+
+        if res.rows_affected() == 0 {
+            return Err(DbError::SetBlockFail);
+        }
+
+        Ok(())
+    }
+
+    async fn add_tx(&self, metadata: &TxMetadata) -> Result<(), DbError> {
+        sqlx::query(
+            "INSERT INTO transactions (
+                tx_hash, 
+                block_number, 
+                from_address, 
+                block_timestamp
+            ) VALUES ($1, $2, $3, $4)
+            ON CONFLICT (tx_hash) DO NOTHING",
```

Comment:
I like that this makes the DB updates idempotent, which should make us more resilient to failures/crashes in the middle of processing a block. But with this approach the errors get swallowed- would prefer us to catch the conflict in the application code and log some sort of warning, so we can have alerts if we keep seeing this.

Reply:
We could refactor to catch those conflicts. The reason I've implemented this way was to keep the code as simple as possible, there may be cases where multiple events (e.g., proof delivered) are batched within the same tx. Surely, it can be refactor if you think getting out that warning is worth though

Reply:
Added with [2b0307a](https://github.com/boundless-xyz/boundless/pull/560/commits/2b0307a3c3ad61a430f820f6e157d8133d8fa049)

Reply:
At the end that solution was not good enough, so I've reverted to just ignore all the potential conflicts